### DISCUSSION TITLE
Class invariant contracts?

### DISCUSSION THREAD

**Discussion by macrologist** [coalton-lang/coalton#953] [Ideas and Proposals]
Typeclasses are sometimes defined with implicit assumptions about invariants that methods of those classes must satisfy for every class instance. E.g. monad laws, ordering laws, etc. 

Could coalton support an extension that allows for the expression of these "invariance contracts" at the typeclass definition site, contracts which would be checked post-compile for each instance, signalling some kind of "instance invariant violation errror"?

**Comment by macrologist** [comment]
Or, even just some access to compilation in the form of "compilation hooks" that could be installed to test those kinds of invariants in a more ad-hoc fashion.

**Reply by stylewarning** [reply]
Do you mean testing a la quickcheck?

**Reply by macrologist** [reply]
Or there's syntax at the instance definition site to say how  you want to create the data that will be tested, and then maybe some compiler feature to complain when those data are missing on a class for which a contract is available. 

Quickeckable data types might be included in such a scheme.
